crypto/punycode.c

```diff
@@ -252,35 +253,36 @@ int ossl_a2ulabel(const char *in, char *out, size_t *outlen)
      * If it does not start with xn--,    it becomes U-label as is.
      * Otherwise we try to decode it.
      */
-    char *outptr = out;
+    char nullbuf[1];
     const char *inptr = in;
-    size_t size = 0, maxsize;
+    size_t maxsize;
     int result = 1;
-    unsigned int i, j;
+    unsigned int i;
     unsigned int buf[LABEL_BUF_SIZE];      /* It's a hostname */
+    WPACKET pkt;
+
+    /* Internal API, so should not fail */
+    if (!ossl_assert(outlen != NULL))
+        return -1;
 
     if (out == NULL) {
+        out = nullbuf;
         result = 0;
-        maxsize = 0;
+        maxsize = sizeof(nullbuf);
     } else {
         maxsize = *outlen;
     }
 
-#define PUSHC(c)                    \
-    do                              \
-        if (size++ < maxsize)       \
-            *outptr++ = c;          \
-        else                        \
-            result = 0;             \
-    while (0)
+    if (!WPACKET_init_static_len(&pkt, (unsigned char *)out, maxsize, 0))
```

Comment:
We can change the semantics of ossl_a2ulabel to count the necessary bytes only in the out==NULL case. Also in that case we could make it return success as there is no real failure. The current semantics is quite awkward.

Reply:
Exactly.....

It's never called with a NULL output buffer.

Maybe this entire thing needs a rethink.  The returned size doesn't appear to be used so maybe the length return can be removed entirely.